### PR TITLE
Add no labels and no milestones and no assignees filters

### LINKED ISSUE #3407: Add "no label" and "no milestone" filters
- Gitea version (or commit ref):
- Git version: N/A
- Operating system: N/A
- Database (use `[x]`): N/A
- Can you reproduce the bug at https://try.gitea.io:
  - [x] Yes (provide example URL) - URL pending
- Log gist: N/A

## Description

I am a project manager triaging issues. I want to filter for all issues without any label, and for all issues without any milestone, so that I know which issues I haven't yet triaged.

Related Gogs issues:
* https://github.com/gogits/gogs/issues/2622
* https://github.com/gogits/gogs/issues/1833
* https://github.com/gogits/gogs/issues/5014

## Suggested solution

* To the label filter list:
  * Add "No label" option
  * Rename "No selected label" to "All labels"
* To the milestone filter list:
  * Add "No milestone" option
  * Rename "No selected milestone" to "All milestones"

### PR DESCRIPTION
**Pull request by tyroneyeh**
Hello,
Sorry, this is my first time to open a pull request, sorry for any mistakes
For Closes #3407 issue

### CODE REVIEW

File: models/issues/issue.go [hunk 1]
```diff
@@ -1663,28 +1701,41 @@ func getIssueStatsChunk(opts *IssueStatsOptions, issueIDs []int64) (*IssueStats,
...
+					} else if labelID < 0 {
 						sess.Where("issue.id NOT IN (SELECT issue_id FROM issue_label WHERE label_id = ?)", -labelID)
 					}
 				}
 			}
 		}
 
-		if opts.MilestoneID > 0 {
-			sess.And("issue.milestone_id = ?", opts.MilestoneID)
+		var mileIDs, noMileIDs []int64
```
Comment (6543) [inline models/issues/issue.go]: we should use a xorm.Builder and not select by In(BigSlice)

File: models/issues/issue.go [hunk 2]
```diff
@@ -1440,9 +1460,27 @@ func issuePullAccessibleRepoCond(repoIDstr string, userID int64, org *organizati
 	return cond
 }
 
-func applyAssigneeCondition(sess *xorm.Session, assigneeID int64) *xorm.Session {
-	return sess.Join("INNER", "issue_assignees", "issue.id = issue_assignees.issue_id").
-		And("issue_assignees.assignee_id = ?", assigneeID)
+func applyAssigneeCondition(sess *xorm.Session, assigneeIDs []int64) *xorm.Session {
```
Comment (6543) [inline models/issues/issue.go]: we should use a xorm.Builder and not select by In(BigSlice)

**Comment by 6543** [comment]
sorry that xorm.Builder is not the easyest thing for beginners - but we have to keep in mind that we had to refactor stuff because it used slices for conditions that just went to big so gitea got slow or just did not work at all anymore

File: models/issues/issue.go [hunk 3]
```diff
@@ -1346,19 +1343,21 @@ func (opts *IssuesOptions) setupSessionNoLimit(sess *xorm.Session) {
...
+	labelsCount := len(opts.LabelIDs)
+	if labelsCount > 0 {
+		noLabelsIDs := make([]int64, 0, labelsCount)
+		hasNoLabel := false
 		for i, labelID := range opts.LabelIDs {
 			if labelID > 0 {
 				sess.Join("INNER", fmt.Sprintf("issue_label il%d", i),
 					fmt.Sprintf("issue.id = il%[1]d.issue_id AND il%[1]d.label_id = %[2]d", i, labelID))
 			} else if labelID < 0 {
 				// sess.Where("issue.id not in (select issue_id from issue_label where label_id = ?)", -labelID)
```
Comment (tyroneyeh) [inline models/issues/issue.go]: image

File: models/issues/issue.go (hunk 1 above)
Comment (tyroneyeh) [reply models/issues/issue.go]: Maybe in 7137988ded1d4ad558992cc01068cc8f9f240080 fixed?

File: models/issues/issue.go (hunk 2 above)
Comment (tyroneyeh) [reply models/issues/issue.go]: Maybe in 7137988ded1d4ad558992cc01068cc8f9f240080 fixed?

File: models/issues/issue.go (hunk 3 above)
Comment (tyroneyeh) [reply models/issues/issue.go]: Do you mean this? It will be slower if you don't use select in!

Comment (6543) [reply models/issues/issue.go]: let em have some time ... I'll have a look once i got time for it - sorry :/

**Comment by tyroneyeh** [comment]
Upload test movie.

https://user-images.githubusercontent.com/1255041/174837864-a4221802-adb2-4519-ab1d-47be5ccff6e4.mov

**Comment by tyroneyeh** [comment]
Hello

I rewrote the code, please review it again

Thanks

File: models/issues/issue.go (hunk 3 above)
Comment (6543) [reply models/issues/issue.go]: Please ping me after 23.

Comment (tyroneyeh) [reply models/issues/issue.go]: @6543 23rd?

File: models/issues/issue.go [hunk 4]
```diff
@@ -1285,6 +1285,9 @@ func (opts *IssuesOptions) setupSessionNoLimit(sess *xorm.Session) {
 
 	if opts.AssigneeID > 0 {
 		applyAssigneeCondition(sess, opts.AssigneeID)
+	} else if opts.AssigneeID == -1 {
+		sess.NotIn("issue.id", builder.Select("issue.id").From("issue").Join("INNER", "issue_assignees", "issue.id = issue_assignees.issue_id").
```
Comment (lunny) [inline models/issues/issue.go]: Maybe
```go
sess.And(builder.NotExists(
      builder.Select("id").From("issue_assignees").
      Where(builder.Expr("issue_assignees.issue_id = issue.id"),
))
```

File: models/issues/issue.go [hunk 5]
```diff
@@ -1335,14 +1342,24 @@ func (opts *IssuesOptions) setupSessionNoLimit(sess *xorm.Session) {
...
 	if opts.LabelIDs != nil {
+		var noLabelsIDs []int64
+		hasNoLabel := false
 		for i, labelID := range opts.LabelIDs {
 			if labelID > 0 {
 				sess.Join("INNER", fmt.Sprintf("issue_label il%d", i),
 					fmt.Sprintf("issue.id = il%[1]d.issue_id AND il%[1]d.label_id = %[2]d", i, labelID))
-			} else {
-				sess.Where("issue.id not in (select issue_id from issue_label where label_id = ?)", -labelID)
+			} else if labelID < 0 {
```
Comment (lunny) [inline models/issues/issue.go:1351]:
```go
sess.And(builder.NotExists(
      builder.Select("id").From("issue_label").
      Where(builder.Expr("issue_label.issue_id = issue.id"),
))
```

File: models/issues/issue.go [hunk 6]
```diff
@@ -1335,14 +1342,24 @@ func (opts *IssuesOptions) setupSessionNoLimit(sess *xorm.Session) {
...
+				// sess.Where("issue.id not in (select issue_id from issue_label where label_id = ?)", -labelID)
+				noLabelsIDs = append(noLabelsIDs, -labelID)
+				hasNoLabel = true
 			}
 		}
+		if hasNoLabel {
+			sess.NotIn("issue.id",
```
Comment (lunny) [inline models/issues/issue.go:1363]:
```go
sess = sess.And(builder.NotExists(
      builder.Select("id").From("issue_label").
      Where(builder.Expr("issue_label.issue_id = issue.id"),
))
```

File: models/issues/issue.go (hunk 3 above)
Comment (lunny) [reply models/issues/issue.go]: This is not right. negative number means this labels should not be contained as conditions.

**Comment by lunny** [comment]
The condition `No label issue` means the issue hasn't been assigned any label.
`Excluded labels` means the issues which doesn't has the special labels.
They are not equal conditions.

File: models/issues/issue.go (hunk 4 above)
Comment (tyroneyeh) [reply models/issues/issue.go]: This modify has error: `Issues: unable to query Issues: Unsupported condition type`

File: models/issues/issue.go (hunk 6 above)
Comment (tyroneyeh) [reply models/issues/issue.go:1363]: This modify in label using alt key is not work

File: models/issues/issue.go [hunk 7]
```diff
@@ -1662,28 +1678,40 @@ func getIssueStatsChunk(opts *IssueStatsOptions, issueIDs []int64) (*IssueStats,
...
 			}
 		}
 
 		if opts.MilestoneID > 0 {
 			sess.And("issue.milestone_id = ?", opts.MilestoneID)
+		} else if opts.MilestoneID == -1 {
+			sess.And("issue.milestone_id = 0")
 		}
 
 		if opts.AssigneeID > 0 {
```
Comment (6543) [inline models/issues/issue.go]:
```suggestion
		if opts.AssigneeID != 0 {
```

File: models/issues/issue.go [hunk 8]
```diff
@@ -1662,28 +1678,40 @@ func getIssueStatsChunk(opts *IssueStatsOptions, issueIDs []int64) (*IssueStats,
...
 			applyAssigneeCondition(sess, opts.AssigneeID)
```
Comment (6543) [inline models/issues/issue.go:1752]: and patch applyAssigneeCondition instead of adding it twice in code

File: models/issues/issue.go [hunk 9]
```diff
@@ -1702,27 +1720,36 @@ func getIssueStatsChunk(opts *IssueStatsOptions, issueIDs []int64) (*IssueStats,
...
+				var noLabelsIDs []int64
+				hasNoLabel := false
 				for i, labelID := range labelIDs {
 					if labelID > 0 {
 						sess.Join("INNER", fmt.Sprintf("issue_label il%d", i),
 							fmt.Sprintf("issue.id = il%[1]d.issue_id AND il%[1]d.label_id = %[2]d", i, labelID))
-					} else {
-						sess.Where("issue.id NOT IN (SELECT issue_id FROM issue_label WHERE label_id = ?)", -labelID)
+					} else if labelID < 0 {
+						// sess.Where("issue.id NOT IN (SELECT issue_id FROM issue_label WHERE label_id = ?)", -labelID)
```
Comment (6543) [inline models/issues/issue.go]:
```suggestion
```

**Review by 6543 — APPROVED** [review]

File: models/issues/issue.go [hunk 10]
```diff
@@ -1449,6 +1462,11 @@ func issuePullAccessibleRepoCond(repoIDstr string, userID int64, org *organizati
 }
 
 func applyAssigneeCondition(sess *xorm.Session, assigneeID int64) *xorm.Session {
+	if assigneeID == -1 {
+		return sess.Join("LEFT", "issue_assignees", "issue.id = issue_assignees.issue_id").
```
Comment (lunny) [inline models/issues/issue.go:1466]: Use `NotExist` is better.

File: models/issues/issue.go [hunk 11]
```diff
@@ -1344,14 +1348,23 @@ func (opts *IssuesOptions) setupSessionNoLimit(sess *xorm.Session) {
 	}
 
 	if opts.LabelIDs != nil {
+		var noLabelsIDs []int64
```
Comment (lunny) [inline models/issues/issue.go:1351]: They are excluded Labels, not no Label. So you still missed real No Label issues.

Comment (6543) [reply models/issues/issue.go:1351]: this is done by set all lables who you dould select as exclude ...
... so it works

the question is if se should use a keyword instead

File: models/issues/issue.go (hunk 10 above)
Comment (6543) [reply models/issues/issue.go:1466]: could you provide an example - never saw an example of `builder.NotExists()`

File: models/issues/issue.go (hunk 11 above)
Comment (lunny) [reply models/issues/issue.go:1351]: If there are hundreds labels in this repository, it's difficult to select labels from the UI?

File: models/issues/issue.go (hunk 10 above)
Comment (lunny) [reply models/issues/issue.go:1466]: > could you provide an example - never saw an example of `builder.NotExists()`

I have done that in the previous comment.

**Pull request closed without merging**